DependencyInjection/NelmioSecurityExtension.php

```diff
@@ -136,4 +147,51 @@ public function load(array $configs, ContainerBuilder $container)
             $container->setParameter('nelmio_security.forced_ssl.hosts', $config['forced_ssl']['hosts']);
         }
     }
+
+    private function buildDirectiveSetDefinition($config, $type)
```

Comment:
I made this pull request nearly a year ago, so you'll have to excuse me if I don't remember all of the details, but here's the gist of it.

If you look at the older code the `ContentSecurityPolicyListener` has a `fromConfig` factory method that is used to build a directive set for both legacy and up-to-date configuration. 

```php
    public static function fromConfig(array $config)
    {
        if (array_key_exists('report', $config) || array_key_exists('enforce', $config)) {
            $enforce = DirectiveSet::fromConfig($config, 'enforce');
            $report = DirectiveSet::fromConfig($config, 'report');
        } else {
            // legacy config
            $directiveSet = DirectiveSet::fromLegacyConfig($config);

            if ((bool) $config['report_only']) {
                $enforce = new DirectiveSet();
                $report = $directiveSet;
            } else {
                $enforce = $directiveSet;
                $report = new DirectiveSet();
            }
        }

        return new self($report, $enforce, (bool) $config['compat_headers'], $config['hosts'], $config['content_types']);
    }
```

I believe it has no place in this listener.

When adding additional dependencies you are forced to build them in the factory method (the same way DirectiveSet instances are built) or pass them around in multiple places (the constructor and the factory method, including all of its usages). This also complicates injection of optional dependencies that are toggled in the bundle configuration. 

I wanted to inject already built directive sets into the constructor, so I essentially moved the code that does so from the listener. Additionally, I moved the logic from `DirectiveSet::fromLegacyConfig` because the extension is more than capable of handling legacy configuration mapping. I should've removed `getLegacyNamesMap` as well and let the extension do all of the dirty glue work with no pollution in the actual directive class.

I was planning on making further incremental improvements and cleanups after my PR, but got discouraged by the lack of interest for this functionality and ended up using my own fork instead.

TL;DR: most of the code is taken from factory methods.